### ISSUE TITLE
unigram_model_trainer stuck on "Extracting frequent sub strings..."

### ISSUE THREAD

**Issue by Talisberg** [google/sentencepiece#641]
Hey,
I'm trying to train the tokenizer and it's stuck on "Extracting frequent sub strings".

I'm using the pip installation for the package, in 2 different machines
1. Windows, i7, 16 cores, 16 RAM
2. linux ubuntu, 16 cores, 120 RAM

args:
```
max_sentence_length: 1000,
vocab_size: 8000,
max_sentencepiece_length: 10,
model_type: UNIGRAM,
num_threads: 16
```

and the same thing has happened in both machines:
1st Test:
a file with 20,000 Sentences.
```
CPU times: user 4.47 s, sys: 84 ms, total: 4.55 s
Wall time: 2.32 s
```

2nd Test
same 20,000 Sentences copied 12 times to the same file
about 240,000 total sentences.

and here it gets stuck on "Extracting frequent sub strings" 
until I just gave up.

so 2 questions:
1. what can I do to improve the running performances
2. under >Fast and lightweight: Segmentation speed is around 50k sentences/sec, and memory footprint is around 6MB.>
what was the environment to get such results?

And Am I missing something?

**Comment by thiennp223** [comment]
For me, just because it takes time to process.

I waited and it runs okay.

**Comment by AdolfVonKleist** [comment]
I have encountered the same issue for some time.  I tracked it down to this line:

- https://github.com/google/sentencepiece/commit/329383b455a5795f3d182159eb0985a3f20f0fa2#diff-51c497698d155a53cbd00dfd1a26018599599affa606d6fd44d663d18d4535aaL116 
   - ` array.push_back(kSentenceBoundary);  // sentence boundary marker.` 

which was removed from:
- https://github.com/google/sentencepiece/blob/7b19d68be0d516d49572a0fa877473df904f90c2/src/unigram_model_trainer.cc#L116

if I understand correctly, the pre-merge version adds an explicit sentence boundary marker for each sentence in the provided training corpus (by default indicated by the newline).  This change effectively treats the entire corpus as a single sentence, and dramatically slows down the training due to the size of the suffix array.  

If you roll back your release to:
 - https://github.com/google/sentencepiece/releases/tag/v0.1.86
 - `pip install sentencepiece==0.1.86` 

you'll see an enormous speedup.  For me, comparing on a small subset of 85k sentences from librispeech, the difference with and without this line was: 12s (with the boundary marker in the loop), and hours (without it).  The 0.1.86 version is what is installed by default by a number of ensemble libraries like espnet.  I'm not sure if there is a perhaps a slight improvement in downstream accuracy from this, but the difference in speed for large corpora is such that it might not be warranted in practice.
